File: 034_SearchForARange.cpp
//1.88%
//My first method
class Solution {
public:
    vector<int> searchRange(vector<int>& nums, int target) {
        int size(nums.size());
		vector<int> res(2);
		int left(0), right(size - 1);
		int mid(-1);
		while(left <= right){
			mid = (left + right) / 2;
			if(nums[mid] > target)
				right = mid - 1;
			else if(nums[mid] < target)
				left = mid + 1;
			else
				break;
		}
		if(left > right){
			res[0] = -1;
			res[1] = -1;
			return res;
		}
		left = mid; right = mid;
		while(left > 0 && nums[left - 1] == target) left--;
		while(right < size - 1 && nums[right + 1] == target) right++;
		res[0] = left;
		res[1] = right;
		return res;	
    }
};

//43.77%
//My second method
class Solution {
public:
    vector<int> searchRange(vector<int>& nums, int target) {
		int size(nums.size());
		vector<int> res(2, -1);
		if(nums.empty()) return res;
		
		//Find out low bound
		int low(-1), up(-1);
		if(nums[0] == target)
			low = 0;
		else if(nums[0] > target)
			return res;
		else{
			int left(0), right(size - 1);
			while(left <= right){
				int mid = (left + right) / 2;
				if(nums[mid] > target)
					right = mid - 1;
				else if(nums[mid] < target)
					left = mid + 1;
				else if(nums[mid - 1] < target){
					low = mid;
					break;
				}else
					right = mid - 1;
			}
		}
		
		//Find up bound
		if(nums[size - 1] == target)
			up = size - 1;
		else if(nums[size - 1] < target)
			return res;
		else{
			int left = 0;
			int right = size - 1;
			while(left <= right){
				int mid = (left + right) / 2;
				if(nums[mid] > target)
					right = mid - 1;
				else if(nums[mid] < target)
					left = mid + 1;
				else if(nums[mid + 1] > target){
					up = mid;
					break;
				}else
					left = mid + 1;
			}
		}
		res[0] = low;
		res[1] = up;
		return res;			
    }
};

//A better solution than mine and told us how to obtain the bound value in a sorted sequence
vector<int> searchRange(int A[], int n, int target) {
    int i = 0, j = n - 1;
    vector<int> ret(2, -1);
    // Search for the left one
    while (i < j)
    {
        int mid = (i + j) /2;
        if (A[mid] < target) i = mid + 1;
        else j = mid;
    }
    if (A[i]!=target) return ret;
    else ret[0] = i;
    
    // Search for the right one
    j = n-1;  // We don't have to set i to 0 the second time.
    while (i < j)
    {
        int mid = (i + j) /2 + 1;	// Make mid biased to the right
        if (A[mid] > target) j = mid - 1;  
        else i = mid;				// So that this won't make the search range stuck.
    }
    ret[1] = j;
    return ret; 
}

//By using std library
vector<int> searchRange(vector<int>& nums, int target) {
    auto bounds = equal_range(nums.begin(), nums.end(), target);
    if (bounds.first == bounds.second)
        return {-1, -1};
    return {bounds.first - nums.begin(), bounds.second - nums.begin() - 1};
}

vector<int> searchRange(vector<int>& nums, int target) {
    int lo = lower_bound(nums.begin(), nums.end(), target) - nums.begin();
    if (lo == nums.size() || nums[lo] != target)
        return {-1, -1};
    int hi = upper_bound(nums.begin(), nums.end(), target) - nums.begin() - 1;
    return {lo, hi};
}

File: 034_SearchForARange.py
# 70.07%
class Solution(object):
    def searchRange(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: List[int]
        """
        res = [-1] * 2
        size = len(nums)
        if size == 0:
            return res
        #Find low bound
        low, up = -1, -1
        if nums[0] > target:
            return res
        elif nums[0] == target:
            low = 0
        else:
            left, right = 0, size - 1
            while left <= right:
                mid = (left + right) / 2
                if nums[mid] > target:
                    right = mid - 1
                elif nums[mid] < target:
                    left = mid + 1
                elif nums[mid - 1] < target:
                    low = mid
                    break
                else:
                    right = mid - 1
        
        #Find up bound
        if nums[size - 1] < target:
            return res
        elif nums[size - 1] == target:
            up = size - 1
        else:
            left, right = 0, size - 1
            while left <= right:
                mid = (left + right) / 2
                if nums[mid] > target:
                    right = mid - 1
                elif nums[mid] < target:
                    left = mid + 1
                elif nums[mid + 1] > target:
                    up = mid
                    break
                else:
                    left = mid + 1
        
        res[0], res[1] = low, up
        return res
	
# A very elegant recursive function	
def searchRange(self, nums, target):
    def search(lo, hi):
        if nums[lo] == target == nums[hi]:
            return [lo, hi]
        if nums[lo] <= target <= nums[hi]:
            mid = (lo + hi) / 2
            l, r = search(lo, mid), search(mid+1, hi)
            return max(l, r) if -1 in l+r else [l[0], r[1]]
        return [-1, -1]
    return search(0, len(nums)-1)
	
# What a genious
def searchRange(self, nums, target):
    def search(n):
        lo, hi = 0, len(nums)
        while lo < hi:
            mid = (lo + hi) / 2
            if nums[mid] >= n:
                hi = mid
            else:
                lo = mid + 1
        return lo
    lo = search(target)
    return [lo, search(target+1)-1] if target in nums[lo:lo+1] else [-1, -1]

def searchRange(self, nums, target):
    lo = bisect.bisect_left(nums, target)
    return [lo, bisect.bisect(nums, target)-1] if target in nums[lo:lo+1] else [-1, -1]
